chains/substrate/events/handlers.go

```diff
@@ -32,3 +34,38 @@ func (eh *SystemUpdateEventHandler) HandleEvents(evts Events, msgChan chan []*me
 	}
 	return nil
 }
+
+type DepositHandler interface {
+	HandleDeposit(sourceID uint8, destID uint8, nonce uint64, resourceID types.ResourceID, calldata []byte, depositType message.TransferType, handlerResponse []byte) (*message.Message, error)
+}
+
+type FungibleTransferEventHandler struct {
+	domainID       uint8
+	depositHandler DepositHandler
+}
+
+func NewFungibleTransferEventHandler(domainID uint8, depositHandler DepositHandler) *FungibleTransferEventHandler {
+	return &FungibleTransferEventHandler{
+		depositHandler: depositHandler,
+		domainID:       domainID,
+	}
+}
+
+func (eh *FungibleTransferEventHandler) HandleEvents(evts Events, msgChan chan []*message.Message) error {
+	domainDeposits := make(map[uint8][]*message.Message)
+	for _, d := range evts.Deposit {
+		func(d Deposit) {
```

Comment:
This was done on core so it had recover() in case the HandleDeposit throwed a panic (which it did if some data was malformed)